src/broadcast/broadcast.rs

```diff
@@ -284,21 +365,88 @@ impl<N: NodeIdT> Broadcast<N> {
+        }
+
         Ok(step.join(self.compute_output(hash)?))
     }
 
     /// Sends an `Echo` message and handles it. Does nothing if we are only an observer.
     fn send_echo(&mut self, p: Proof<Vec<u8>>) -> Result<Step<N>> {
-        self.echo_sent = true;
-        if !self.netinfo.is_validator() {
+       if !self.netinfo.is_validator() {
             return Ok(Step::default());
         }
         let echo_msg = Message::Echo(p.clone());
-        let step: Step<_> = Target::All.message(echo_msg).into();
+        let mut step = Step::default();
+        // TODO: iterator not correctly describing right of our_id. Use cycle method on
+        // iterator to correctly specify.
+        let right = self.netinfo.all_ids().skip(self.netinfo.num_correct()+self.pessimissm_factor);
+
+        // Send `Echo` to all left nodes.
+        if !self.echo_sent_partial {
+            self.echo_sent_partial = true;
+            for id in right {
+                let msg = Target::Node(id.clone()).message(echo_msg.clone());
+                step.messages.push(msg);
+            }
+        }
+        // Send `Echo` to remaining right nodes who haven't sent us `CanDecode`.
+        else {
+            self.echo_sent_full = true;
+            for id in right {
+                if !self.can_decodes.contains_key(id) {
+                    let msg = Target::Node(id.clone()).message(echo_msg.clone());
+                    step.messages.push(msg);
+                }
+            }
+        }
         let our_id = &self.our_id().clone();
         Ok(step.join(self.handle_echo(our_id, p)?))
     }
 
+    /// Sends an `EchoHash` message and handles it. Does nothing if we are only an observer.
+    fn send_echo_hash(&mut self, hash: &Digest) -> Result<Step<N>> {
+        self.echo_hash_sent = true;
+        if !self.netinfo.is_validator() {
+            return Ok(Step::default());
+        }
+        let echo_hash_msg = Message::EchoHash(*hash);
+        let mut step = Step::default();
+        // TODO: iterator not correctly describing left of our_id. Use cycle method on
+        // iterator to correctly specify.
+        let left = self.netinfo.all_ids().take(self.netinfo.num_correct()+self.pessimissm_factor);
+        for id in left {
+            let msg = Target::Node(id.clone()).message(echo_hash_msg.clone());
+            step.messages.push(msg);
+        }
+        let our_id = &self.our_id().clone();
+        Ok(step.join(self.handle_echo_hash(our_id, hash)?))
+    }
+
+    /// Sends a `CanDecode` message and handles it. Does nothing if we are only an observer.
+    fn send_can_decode(&mut self, hash: &Digest) -> Result<Step<N>> {
+        self.can_decode_sent = true;
+        if !self.netinfo.is_validator() {
+            return Ok(Step::default());
+        }
+
+        let can_decode_msg = Message::CanDecode(*hash);
+        let mut step = Step::default();
+
+        for id in self.netinfo.all_ids() {
+            if self.echos.contains_key(id) {
```

Comment:
Should this be negated? `CanDecode` needs to go to everyone who _hasn't_ sent us their `Echo` yet.
(Also, should we filter out ourselves here? Possibly even add a `other_ids` method to `NetworkInfo`?)

Reply:
Yup you are right. Will fix this.